Add HTTPS support for web playlists (M3U, PLS, RA, ASX etc.)

t3chnical1 opened the pull request:
Adds HTTPS support for web playlists (M3U, PLS etc.) HTTP 302 redirections, and various fixes to detections of playlists, as none currently exists.

I've kept the majority of the original code for HTTP (lots of fallback code) and based my modifications on "UpdateChecker.cpp" for handling HTTPS.

This was tested with a list of 60+ different online radio channels, many of which did not previously work.

The current mechanics of handling redirections are a bit dodgy to play with in MPC and would require some doing to change, even though it would probably benefit our handling of Youtube-DL urls also. I kept my modifications simple to avoid the need for major testing.

This is my first official patch, hope it goes well :)

Cheers!

Commit:
Add HTTPS support for web playlists (M3U, PLS etc.) and HTTP 302 redirections, and various fixes to detections of playlists.

Commit:
Remove redirection information. Causes a bug with audio/mpeg content when redirected. Redirection is handled correctly elsewhere.

ale5000-git commented:
Why you remove proxy support?

I'm not related to this project but I want to say this: I think your contribute is appreciated and also I understand that everyone has its own style but I think you should try to keep existing code style like newlines (this also help to easily review the changes).

t3chnical1 commented:
I should have mentioned the Proxy. I tested with a local proxy and saw the calls went through it. Contrary to the CSocket class, it seems that the CInternetSession handles the proxy is automatically.

If anyone has a restrictive proxy they can test on, that is that the internet traffic *must* go through the proxy and can confirm that, it would be appreciated. I am able to set a proxy, but not block all other traffic.

As for the newlines, are you referring to CRLF as end of lines? Can you give an example?

ale5000-git commented on src/mpc-hc/mplayerc.cpp:
```diff
@@ -347,207 +349,212 @@ CStringA GetContentType(CString fn, CAtlList<CString>* redir)
-        } else if (ext == _T(".qtl")) {
-            ct = _T("application/x-quicktimeplayer");
+        }
+        else if (ext == _T(".swf")) {
```
@t3chnical1: For example this line, there is an additional newline.

t3chnical1 replied:
I agree. This got by me. I will be more careful in the future.

ale5000-git commented:
@t3chnical1: Do you mean that if you set a custom proxy in the settings of  Windows it will automatically use it?
I mean a normal manually configured proxy, not a corporate proxy inside the lan.

The registry settings are explained here: https://docs.microsoft.com/en-us/troubleshoot/windows-client/networking/configure-client-proxy-server-settings-by-registry-file

t3chnical1 commented:
@ale5000-git : During my testing (and I just checked it again to make sure), I can confirm that setting a proxy in Windows 10 is handled correctly by CInternetSession. (Start Menu -> Settings -> Network & Internet -> Proxy).
I'm not able to test older versions of Windows as I have none readily available.

Commit:
Merge branch 'develop' of https://github.com/clsid2/mpc-hc into develop

clsid2 commented on src/mpc-hc/mplayerc.cpp:
```diff
@@ -347,207 +349,212 @@ CStringA GetContentType(CString fn, CAtlList<CString>* redir)
+        CString osVersionStr;
+        osVersionStr.Format(_T("Windows %1u.%1u"), osVersion.dwMajorVersion, osVersion.dwMinorVersion);
+
+#if !defined(_WIN64)
```
Use just "MPC-HC" as user agent.

t3chnical1 replied:
Done. While coding it, I was wondering about implementing a configurable global variable for the User-Agent. Some servers might only respond to web browsers...

ale5000-git replied:
@clsid2: What about an advanced options "Use browser User-Agent" with true/false that when true use an User-Agent similar to the one of youtube-dl?
This: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3768.2 Safari/537.36`

To note: they identify as Windows 10 also on other versions of Windows.

clsid2 commented on src/mpc-hc/mplayerc.cpp:
```diff
@@ -347,207 +349,212 @@ CStringA GetContentType(CString fn, CAtlList<CString>* redir)
-                    }
-                }
-            }
+        CHttpFile* versionFile = NULL;
```
Think of better variable name than "versionFile"

clsid2 commented on src/mpc-hc/mplayerc.cpp:
```diff
@@ -347,207 +349,212 @@ CStringA GetContentType(CString fn, CAtlList<CString>* redir)
+                    cnt = "audio/x-pn-realaudio";
+                    exit = true;
                 }
                 if (!strncmp((LPCSTR)str, ".RMF", 4)) {
```
Change these into "else if"

Commit:
Set Internet Option Connect Timeout to 10 secs (prev 45s) and address code review comments

t3chnical1 commented:
I have addressed the code review comments. While testing I had the chance to encounter a non-responding server (downed server). I noticed the default timeout when opening a connection was 45 secs which caused Windows to think the app was not responding. I set a 10 sec timeout (which is still find very high). When that timeout occurs, we still have another 30sec or so timeout that eventually expires because the URL is still sent to the LAVFilters/FFMPEG. The status bar indicates "Opening..." during the last 30 secs and the app responds normally by stating the connection failed.

I'm open to comments. And thanks for the review!

clsid2 commented:
A 5 sec timeout is pretty standard I think.

Add a "BOOL url_fail" and set that to true when there is a failure condition for URL connection. For example also for 404/403 status.

I will think of a nice way to propagate that to the calling code later to avoid opening the URL again, and to show a proper status message. I might replace it with a CString containing the error message, so use a separate `url_fail=true` for each unique error condition, so it can easily be replaced later. Stuff needs to be translatable, so I prefer doing it in a separate patch.

Commit:
Handle return HTTP Status codes and lower connection open timeout to 5 secs.

t3chnical1 commented:
Timeout is now 5 secs and I added as you suggested local handling for the HTTP codes without changing the function signature.
